internal/controller/controller.go

```diff
@@ -228,6 +228,8 @@ func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
 	// happens only after three sync intervals (900 seconds).
 	go wait.Until(c.periodicCleanup, 907*time.Second, stopCh)
 
+	go wait.Until(c.ensureAgentsState, 300*time.Second, stopCh)
```

Comment:
Theoretically possible, I added a mutex in my latest commit to ensure this won't happen.

Reply:
`EnsureAgentsState` is always called after the port creation within the `ProvisionPort` so it should always update the address accordingly afterwards or am I missing something? 
The only problem I noticed was that an old result of `GetPorts` could be used for the `UpdateAgents` call.

Reply:
Okay, nevermind, I think I was confused.

I thought the workqueue was parallelized, but if it were, everything would go up in flames anyway because I think none of our data structures are actually thread safe.

So the condition I was thinking of (`EnsureAgentsState` vs. `SyncService`) can't happen and you can remove the mutex again…